File: frontend/src/api/policyBuilder.ts
import { apiClient } from './client'

export interface NewObjectSpec {
  name: string
  object_kind: 'address' | 'service'
  address_type?: 'ip-mask' | 'ip-range' | 'fqdn'
  ip_address?: string
  protocol?: 'tcp' | 'udp'
  port?: string
  description?: string
}

export interface NewPolicyRow {
  row_index: number
  rule_name: string
  rule_action: string
  disabled: boolean
  from_zone: string
  source: string
  source_user: string
  to_zone: string
  destination: string
  service: string
  application: string
  description: string
  log_end: string
  log_setting: string
}

export interface MoveTarget {
  position: 'top' | 'bottom' | 'before' | 'after'
  reference_policy_id: number | null
}

export interface ObjectGapItem {
  name: string
  object_kind: 'address' | 'service'
  referenced_by_rule_names: string[]
}

export interface GeneratedCommand {
  row_index: number
  kind: 'object' | 'policy' | 'move'
  command: string | null
  error: string | null
  counts: Record<string, number> | null
}

export interface InsertionConflict {
  rule_name: string
  conflict_type: 'blocking' | 'shadowing'
  conflicting_policy_id: number
  conflicting_policy_name: string
  reason: string
}

export interface PreviewRow {
  id: number
  rule_name: string
  action: string
  seq: number | null
  is_new: boolean
}

export interface BulkPolicyPlanResponse {
  missing_objects: ObjectGapItem[]
  object_commands: GeneratedCommand[]
  policy_commands: GeneratedCommand[]
  move_commands: GeneratedCommand[]
  modify_commands: GeneratedCommand[]
  delete_commands: GeneratedCommand[]
  conflicts: InsertionConflict[]
  preview_before: PreviewRow[]
  preview_after: PreviewRow[]
  warnings: string[]
}

export interface PreviewPolicyRow {
  id: number
  device_id: number
  rule_name: string
  action: string
  seq: number | null
  source: string
  destination: string
  service: string
  application: string | null
  from_zone: string | null
  to_zone: string | null
  user: string | null
  description: string | null
  log_setting: string | null
  enable: boolean | null
  security_profile: string | null
  category: string | null
  last_hit_date: string | null
  hit_count: number | null
  is_active: boolean
  last_seen_at: string | null
  vsys: string | null
  pending_status: 'new' | 'modified' | 'deleted' | 'moved' | null
}

export type PendingChangeType = 'create' | 'new_object' | 'modify' | 'delete' | 'move'

export interface PendingPolicyChange {
  id: number
  device_id: number
  change_type: PendingChangeType
  target_policy_id: number | null
  client_key: string
  payload: Record<string, unknown>
  created_by_user_id: number | null
  created_at: string
}

export const listPendingChanges = async (deviceId: number): Promise<PendingPolicyChange[]> => {
  const res = await apiClient.get<PendingPolicyChange[]>(`/policy-builder/${deviceId}/pending-changes`)
  return res.data
}

export const addPendingChange = async (
  deviceId: number,
  payload: { change_type: PendingChangeType; target_policy_id?: number | null; client_key: string; payload: Record<string, unknown> },
): Promise<PendingPolicyChange> => {
  const res = await apiClient.post<PendingPolicyChange>(`/policy-builder/${deviceId}/pending-changes`, payload)
  return res.data
}

export const updatePendingChange = async (
  deviceId: number,
  changeId: number,
  payload: Record<string, unknown>,
): Promise<PendingPolicyChange> => {
  const res = await apiClient.patch<PendingPolicyChange>(`/policy-builder/${deviceId}/pending-changes/${changeId}`, { payload })
  return res.data
}

export const removePendingChange = async (deviceId: number, changeId: number): Promise<void> => {
  await apiClient.delete(`/policy-builder/${deviceId}/pending-changes/${changeId}`)
}

export const getPreviewOrder = async (deviceId: number): Promise<PreviewPolicyRow[]> => {
  const res = await apiClient.get<PreviewPolicyRow[]>(`/policy-builder/${deviceId}/preview-order`)
  return res.data
}

export const clearPendingChanges = async (deviceId: number): Promise<void> => {
  await apiClient.delete(`/policy-builder/${deviceId}/pending-changes`)
}

/**
 * 신규 정책 생성(create) 변경사항이 취소되면, 그 정책만 참조하던 신규 오브젝트(new_object)
 * 변경사항이 고아로 남을 수 있다 — 남은 create 변경사항 중 어느 것도 이름을 참조하지 않는
 * new_object만 골라 정리한다. `remainingChanges`는 삭제 대상을 제외한, 정리 시점의 전체 목록.
 */
export const cleanupOrphanNewObjects = async (deviceId: number, remainingChanges: PendingPolicyChange[]): Promise<void> => {
  const referencedNames = new Set<string>()
  for (const c of remainingChanges) {
    if (c.change_type !== 'create') continue
    for (const field of ['source', 'destination', 'service']) {
      const raw = String(c.payload[field] ?? '')
      for (const token of raw.split(',')) {
        const name = token.trim().replace(/^"|"$/g, '')
        if (name) referencedNames.add(name)
      }
    }
  }
  const orphaned = remainingChanges.filter(
    (c) => c.change_type === 'new_object' && !referencedNames.has(String(c.payload.name ?? ''))
  )
  await Promise.all(orphaned.map((c) => removePendingChange(deviceId, c.id)))
}

export const checkObjectGaps = async (deviceId: number, newPolicies: NewPolicyRow[]): Promise<ObjectGapItem[]> => {
  const res = await apiClient.post(`/policy-builder/${deviceId}/object-gaps`, { new_policies: newPolicies })
  return res.data.missing_objects
}

export const planBulkPolicy = async (
  deviceId: number,
  payload: { vsys?: string } = {},
): Promise<BulkPolicyPlanResponse> => {
  const res = await apiClient.post<BulkPolicyPlanResponse>(`/policy-builder/${deviceId}/plan`, payload)
  return res.data
}

export function newPolicyRowDefaults(rowIndex: number): NewPolicyRow {
  return {
    row_index: rowIndex,
    rule_name: '',
    rule_action: 'allow',
    disabled: false,
    from_zone: '',
    source: '',
    source_user: '',
    to_zone: '',
    destination: '',
    service: '',
    application: '',
    description: '',
    log_end: '',
    log_setting: '',
  }
}

File: backend/app/schemas/policy_builder.py
from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel


class NewObjectSpec(BaseModel):
    """사용자가 입력한, 아직 장비에 없는 신규 오브젝트의 실제 값."""
    name: str
    object_kind: Literal["address", "service"]
    address_type: Optional[Literal["ip-mask", "ip-range", "fqdn"]] = None
    ip_address: Optional[str] = None
    protocol: Optional[Literal["tcp", "udp"]] = None
    port: Optional[str] = None
    description: str = ""


class NewPolicyRow(BaseModel):
    """붙여넣기로 입력받은 신규 정책 1건."""
    row_index: int
    rule_name: str
    rule_action: str = "allow"
    disabled: bool = False
    from_zone: str = ""
    source: str = ""
    source_user: str = ""
    to_zone: str = ""
    destination: str = ""
    service: str = ""
    application: str = ""
    description: str = ""
    log_end: str = ""
    log_setting: str = ""


class MoveTarget(BaseModel):
    """신규 정책들을 배치할 목표 위치. reference_policy_id는 DB에 이미 존재하는 정책만 허용."""
    position: Literal["top", "bottom", "before", "after"]
    reference_policy_id: Optional[int] = None


class ObjectGapItem(BaseModel):
    name: str
    object_kind: Literal["address", "service"]
    referenced_by_rule_names: List[str]


class ObjectGapCheckRequest(BaseModel):
    new_policies: List[NewPolicyRow]


class ObjectGapCheckResponse(BaseModel):
    missing_objects: List[ObjectGapItem]


class BulkPolicyPlanRequest(BaseModel):
    """Policies 편집모드의 '대기중 변경사항'을 모두 모아 CLI를 생성할 때 쓰는 요청. vsys 외에는 DB의
    PendingPolicyChange 레코드에서 읽어온다."""
    vsys: Optional[str] = None


class FieldDiff(BaseModel):
    """그리드 셀 편집 시 원본 값과 편집된 값을 비교해 계산된, 필드 하나의 추가/삭제된 토큰."""
    added: List[str] = []
    removed: List[str] = []


class GeneratedCommand(BaseModel):
    row_index: int
    kind: Literal["object", "policy", "move", "modify", "delete"]
    command: Optional[str] = None
    error: Optional[str] = None
    counts: Optional[Dict[str, int]] = None


class InsertionConflict(BaseModel):
    rule_name: str
    conflict_type: Literal["blocking", "shadowing"]
    conflicting_policy_id: int
    conflicting_policy_name: str
    reason: str


class PreviewRow(BaseModel):
    id: int
    rule_name: str
    action: str
    seq: Optional[int] = None
    is_new: bool = False


class PreviewPolicyRow(BaseModel):
    """편집모드 그리드에 표시할, 대기중 변경사항(생성/수정/삭제/이동)을 모두 적용한 뒤의 정책 1행.

    신규 생성행은 `id`가 음수(-pending_change.id)이며 `seq`/`last_hit_date` 등 DB에서만
    채워지는 필드는 비어 있다."""
    id: int
    device_id: int
    rule_name: str
    action: str = "allow"
    seq: Optional[int] = None
    source: str = ""
    destination: str = ""
    service: str = ""
    application: Optional[str] = None
    from_zone: Optional[str] = None
    to_zone: Optional[str] = None
    user: Optional[str] = None
    description: Optional[str] = None
    log_setting: Optional[str] = None
    enable: Optional[bool] = None
    security_profile: Optional[str] = None
    category: Optional[str] = None
    last_hit_date: Optional[datetime] = None
    hit_count: Optional[int] = None
    is_active: bool = True
    last_seen_at: Optional[datetime] = None
    vsys: Optional[str] = None
    pending_status: Optional[Literal["new", "modified", "deleted", "moved"]] = None


class BulkPolicyPlanResponse(BaseModel):
    missing_objects: List[ObjectGapItem]
    object_commands: List[GeneratedCommand]
    policy_commands: List[GeneratedCommand]
    move_commands: List[GeneratedCommand]
    modify_commands: List[GeneratedCommand] = []
    delete_commands: List[GeneratedCommand] = []
    conflicts: List[InsertionConflict]
    preview_before: List[PreviewRow]
    preview_after: List[PreviewRow]
    warnings: List[str]
